src/tools/database/elasticsearch.ts

```diff
@@ -0,0 +1,198 @@
+          propertyName: "connection.cloud, connection.node, connection.nodes",
+        },
+      ]);
+    }
+  }
+
+  @Cache()
+  protected get client(): Client {
+    try {
+      return new Client(this.options.connection);
+    } catch (error) {
+      throw new ToolError(`Unable to connect to ElasticSearch: ${error}`, [], {
+        isRetryable: false,
+        isFatal: true,
+      });
+    }
+  }
+
+  protected async _run(
+    input: ToolInput<this>,
+    _options?: ToolRunOptions,
+  ): Promise<JSONToolOutput<any>> {
+    if (input.action === "LIST_INDICES") {
+      const indices = await this.listIndices(_options?.signal);
+      return new JSONToolOutput(indices);
+    } else if (input.action === "GET_INDEX_DETAILS") {
+      const indexDetails = await this.getIndexDetails(input, _options?.signal);
+      return new JSONToolOutput(indexDetails);
+    } else if (input.action === "SEARCH") {
+      const response = await this.search(input, _options?.signal);
+      if (response.aggregations) {
+        return new JSONToolOutput(response.aggregations);
+      } else {
+        return new JSONToolOutput(response.hits.hits.map((hit: SearchHit) => hit._source));
+      }
+    } else {
+      throw new ToolError("Invalid action specified.");
+    }
+  }
+
+  private async listIndices(signal?: AbortSignal): Promise<CatIndicesResponse> {
+    const response = await this.client.cat.indices(
+      {
+        expand_wildcards: "open",
+        h: "index",
+        format: "json",
+      },
+      { signal: signal },
+    );
+    return response
+      .filter((record) => record.index && !record.index.startsWith(".")) // Exclude system indices
+      .map((record) => ({ index: record.index }));
+  }
+
+  private async getIndexDetails(
+    input: ToolInput<this>,
+    signal?: AbortSignal,
+  ): Promise<IndicesGetMappingResponse> {
+    if (!input.indexName) {
+      throw new ToolError("Index name is required for GET_INDEX_DETAILS action.");
+    }
+    return await this.client.indices.getMapping(
+      {
+        index: input.indexName,
+      },
+      { signal: signal },
+    );
+  }
+
+  private async search(input: ToolInput<this>, signal?: AbortSignal): Promise<SearchResponse> {
+    if (!input.indexName || !input.query) {
+      throw new ToolError("Both index name and query are required for SEARCH action.");
+    }
+    let parsedQuery;
+    try {
+      parsedQuery = JSON.parse(input.query);
+    } catch {
+      throw new ToolError(`Invalid JSON format for query`);
+    }
+
+    const searchBody: SearchRequest = {
+      ...parsedQuery,
+      from: parsedQuery.from || input.start,
+      size: parsedQuery.size || input.size,
+    };
+
+    return await this.client.search(
+      {
+        index: input.indexName,
+        body: searchBody,
+      },
+      { signal: signal },
+    );
+  }
+
+  loadSnapshot({ ...snapshot }: ReturnType<typeof this.createSnapshot>): void {
```

Comment:
`loadSnapshot(snapshot: ReturnType<typeof this.createSnapshot>) {`

Reply:
or you can remove the whole thing

Reply:
removed